### ISSUE TITLE
UnknownExistentialVariable compiler bug triggered by delaying a higher-rank function

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#2354]
Self-contained example:
```
f : (forall a . a -> a) -> Nat
f id = id 0

x = 'f
```
Results in: 
```  
UnknownExistentialVariable:
  unsolved
```

Seems like any higher rank function will do it. (Like try `x = ‘Scope.run`)

Not sure the urgency on this but it does make it impossible to link to higher rank functions in the docs. I’m hoping it’s just something easy to fix.

**Comment by ceedubs** [comment]
I just ran into this, and the fact that you don't get a better error message with a line number definitely makes it rough. You start hunting through your code for where you might have made a bug and it isn't obvious that what you are doing is fine and the compiler is just confused.

**Comment by dolio** [comment]
I took another look at this.

I think I found a bug in in `instantiateL` Essentially, in the case:

    v <= forall a. T

It's supposed to proceed to:

    v <= T[a := fresh]

But the original v is getting shadowed, so instead it does:

    let v = fresh in v <= T[a := v]

When I fix this, what I get is a proper type error (instead of an internal compiler error). My question is: is it _supposed_ to be a type error? It seems like the algorithm forces this, because it's trying to find a HM-type for `f` (because of a lack of annotations), but none exists. But, I wasn't anticipating this case failing because it seems so simple.
